specification/dartLangSpec.tex

```diff
@@ -11716,18 +11716,43 @@ \subsection{Function Expressions}
 }
 
 \LMHash{}%
-We say that $S$ is the
-\IndexCustom{future type derived from a type}{type!future type derived from}
-$T$ in the following cases, using the first applicable case:
+We say that a type $T$
+\IndexCustom{derives a future type}{type!derives a future type}
+$F$ in the following cases, using the first applicable case:
 
+%% TODO(eernst): Note that `X extends X?` can create an infinite loop.
+%% We will probably make that kind of bound an error.
 \begin{itemize}
-\item $T$ implements $S$
-  (\ref{interfaceSuperinterfaces}),
-  and there is a $U$ such that $S$ is \code{Future<$U$>}.
-\item $T$ is $S$ bounded
-  (\ref{bindingActualsToFormals}),
-  and there is a $U$ such that
-  $S$ is \code{FutureOr<$U$>}, \code{Future<$U$>?}, or \code{FutureOr<$U$>?}.
+\item
+  %% TODO(eernst): Come mixin classes and extension types: add them.
+  If $T$ is a type which is introduced by
+  a class, mixin, or enum declaration,
+  and if $T$ or a direct or indirect superinterface
+  (\ref{interfaceSuperinterfaces})
+  of $T$ is \code{Future<$U$>} for some $U$,
+  then $T$ derives the future type \code{Future<$U$>}.
+\item
+  If $T$ is the type \code{FutureOr<$U$>} for some $U$,
+  then $T$ derives the future type \code{FutureOr<$U$>}.
+\item
+  If $T$ is \code{$S$?} for some $S$, and
+  $S$ derives the future type $F$,
+  then $T$ derives the future type \code{$F$?}.
+\item
+  If $T$ is a type variable with bound $B$, and
+  $B$ derives the future type $F$,
+  then $T$ derives the future type $F$.
+\item
+  If $T$ is of the form \code{$X$\,\&\,$S$} then:
```

Comment:
We currently only use `"derives a future type" from *flatten* on the promotion type of an `X&B`, which means that `B` cannot itself be a promoted type variable. So this case should be unreachable.

Reply:
The more I think about that rule, the less I like it. (Seemed like a great idea at the time, etc.)

If we have a promoted type variable which doesn't derive a future type, and the type variable itself does (so its bound does), then someone has actively type-checked their way _out of_ deriving a future type.

That means either promoting to `Null` or `Never` (unlikely) or to the non-future branch of a `FutureOr`, in which case awaiting that should probably not go back and use the `FutureOr` type anyway.

Say, having an `X extends FutureOr<num>` and doing `x is int` to promote to `int` and then awaiting that will give something of type `num`. 
I think it would be fine to use `int` instead, and not have a fallback when flattening a promoted type variable, where the promotion doesn't derive a future type.
(Yes, you shouldn't be awaiting an `int`, but it is currently a valid way to insert a delay. And it might still happen in an async return.)